source/adapters/level_zero/v2/latency_tracker.hpp

```diff
@@ -0,0 +1,104 @@
+//===--------- ur_latency_tracker.cpp - common ---------------------------===//
+//
+// Copyright (C) 2024 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <chrono>
+#include <cstdint>
+#include <limits>
+#include <memory>
+
+#include "logger/ur_logger.hpp"
+
+namespace v2 {
+
+static inline bool trackLatency = []() {
+  return std::getenv("UR_ENABLE_LATENCY_TRACKING") != nullptr;
+}();
+
+class rolling_stats {
+public:
+  rolling_stats(const char *name) : name(name) {}
+
+  ~rolling_stats() {
+    if (trackLatency) {
+      logger::info("[{}] average latency: {}ns", name, estimate());
+      logger::info("[{}] number of samples: {}", name, count());
+    }
+  }
+
+  // track latency by taking the value of duration directly.
+  void trackValue(double value) {
```

Comment:
This needs to be thread-safe.

Reply:
We could make it thread safe but if we use some atomic the value will be less precise I guess so I'm not sure if it's worth it - for now, we don't have any lock-free functions where we could use it anyway.

Reply:
ok, let's leave it as is, but we will need to be careful to remember to use per_thread variables where that matters.